File: src/main/java/twp/tools/Logging.java
package twp.tools;

import arc.*;
import arc.func.*;
import arc.util.*;
import org.javacord.api.entity.channel.TextChannel;
import org.javacord.api.entity.message.Message;
import org.junit.platform.commons.util.*;
import twp.*;
import twp.database.*;

import java.io.*;
import java.sql.*;
import java.text.*;
import java.util.concurrent.CompletableFuture;

import static twp.Main.*;

// Serializing errors and sending messages to cmd is performed from here
public class Logging {
    static final String outDir = Global.dir + "/errors/";

    public static void main(String[] args){
        log("hello");
        log("hello");
    }

    public static CompletableFuture<Message> sendDiscordMessage(TextChannel ch, String key, Object ...args) {
         return ch.sendMessage(translate(key, args));
    }

    public static String translate(String key, Object ...args) {
        return Text.cleanColors(Text.format(bundle.getDefault(key), args));
    }

    public static void info(String key, Object ...args) {
        Log.info(translate(key, args));
    }

    public static void log(String message) {
        log(new RuntimeException(message));
    }

    public static void sendMessage(String key, Object ...args) {
        for(PD pd : db.online.values()) {
            pd.sendServerMessage(key, args);
        }
    }

    public static void log(Runnable run) {
        try {
            run.run();
        } catch (Exception e) {
            log(e);
        }
    }

    public static void log(Throwable t) {
        String ex = ExceptionUtils.readStackTrace(t);
        SimpleDateFormat formatter= new SimpleDateFormat("yyyy-MM-dd 'at' HH-mm-ss-SSS z");
        Date date = new Date(System.currentTimeMillis());
        File f = new File(outDir+formatter.format(date));
        t.printStackTrace();
        try {
            Json.makeFullPath(f.getAbsolutePath());
            f.createNewFile();
            PrintWriter out = new PrintWriter(f.getAbsolutePath());

            out.println(ex);
            out.close();
        } catch(IOException e) { e.printStackTrace();}

    }

    public static <T> void on(Class<T> event, Cons<T> cons) {
        Events.on(event, e -> {
            try{
                cons.get(e);
            } catch(Exception ex) {
                log(ex);
            }
        });
    }

    public static <T> void run(Object event, Runnable listener) {
        Events.run(event, () -> {
            try{
                listener.run();
            } catch(Exception ex) {
                log(ex);
            }
        });
    }
}

File: src/main/kotlin/twp/tools/Logging.kt
package twp.tools

import arc.Events
import arc.func.Cons
import arc.util.Log
import java.io.File
import java.io.IOException
import kotlin.jvm.JvmStatic
import java.text.SimpleDateFormat
import org.javacord.api.entity.channel.TextChannel
import java.util.concurrent.CompletableFuture
import java.lang.RuntimeException
import java.lang.Runnable
import org.junit.platform.commons.util.ExceptionUtils
import java.io.PrintWriter
import java.io.FileOutputStream
import org.javacord.api.entity.message.Message
import twp.Global
import twp.Main
import java.lang.Exception
import java.sql.Date

// Serializing errors and sending messages to cmd is performed from here
object Logging {
    const val outDir = Global.dir + "/errors/"
    var formatterDate = SimpleDateFormat("yyyy-MM-dd z")
    var formatterTime = SimpleDateFormat("[HH-mm-ss-SSS]")
    @JvmStatic
    fun main(args: Array<String>) {
        log("hello")
        log("hello")
    }

    @JvmStatic
    fun sendDiscordMessage(ch: TextChannel, key: String?, vararg args: Any?): CompletableFuture<Message> {
        return ch.sendMessage(translate(key, *args))
    }

    @JvmStatic
    fun translate(key: String?, vararg args: Any?): String? {
        return Text.cleanColors(Text.format(Main.bundle.getDefault(key), *args))
    }

    @JvmStatic
    fun info(key: String?, vararg args: Any?) {
        Log.info(translate(key, *args))
    }

    fun log(message: String?) {
        log(RuntimeException(message))
    }

    fun sendMessage(key: String?, vararg args: Any?) {
        for (pd in Main.db!!.online.values) {
            pd.sendServerMessage(key, *args)
        }
    }

    fun log(run: Runnable) {
        try {
            run.run()
        } catch (e: Exception) {
            log(e)
        }
    }

    @JvmStatic
    fun log(t: Throwable) {
        val ex = ExceptionUtils.readStackTrace(t)
        val date = Date(System.currentTimeMillis())
        val f = File(outDir + formatterDate.format(date))
        t.printStackTrace()
        try {
            Json.makeFullPath(f.absolutePath)
            if (!f.exists()) {
                f.createNewFile()
            }
            val out = PrintWriter(FileOutputStream(f, true))
            out.println(formatterTime.format(date))
            out.println(ex)
            out.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun <T> on(event: Class<T>?, cons: Cons<T>) {
        Events.on(event) { e: T ->
            try {
                cons[e]
            } catch (ex: Exception) {
                log(ex)
            }
        }
    }

    fun run(event: Any?, listener: Runnable) {
        Events.run(event) {
            try {
                listener.run()
            } catch (ex: Exception) {
                log(ex)
            }
        }
    }
}